File: R/tagtools/R/dive_stats.R
#' Compute summary statistics for dives or flights
#' 
#' Given a depth/altitude profile and a series of dive/flight start and end times,
#' compute summary dive statistics. 
#' 
#' In addition to the maximum excursion and duration, \code{dive_stats} divides each excursion into three phases:
#' "to" (descent for dives, ascent for flights), "from" (ascent for dives, descent for flights), and "destination". 
#' The "destination" (bottom for dives and top for flights) 
#' phase of the excursion is identified using a "proportion of maximum depth/altitude" method,
#' whereby for example the bottom phase of a dive lasts from the first to the last time the depth exceeds a stated proportion of the maximum depth.
#' Average vertical velocity is computed for the to and from phases using a simple method: total depth/altitude change divided by total time.
#' If an angular data variable is also supplied (for example, pitch, roll or heading),
#' then the circular mean (computed via \code{\link[CircStats]{circ.mean}}) and variance (computed via \code{\link[CircStats]{circ.disp}} and reporting the \code{var} output)
#' are also computed for each dive phase and the dive as a whole.
#' 
#' @param P Depth data. A vector (or one-column matrix), or a tag sensor data list.
#' @param X (optional) Another data stream (as a vector (or a one-column matrix) or a tag sensor data list) for which to compute mean and variability. If \code{angular} is TRUE, interpreted as angular data (for example pitch, roll, or heading) and means and variances are computed accordingly. 
#'  The unit of measure must be radians (NOT degrees). Currently, \code{X} must be regularly sampled.
#' @param dive_cues A two-column data frame or matrix with dive/flight start times in the first column and dive/flight end times in the second. May be obtained from \code{\link{find_dives}}. Units should be seconds since start of tag recording.
#' @param sampling_rate (optional and ignored if \code{P} or \code{X} are tag sensor data lists) Sampling rate of \code{P} (and \code{X}, if \code{X} is given). If omitted, then input data must be sensor data lists. If one value is given and both \code{P} and \code{X} are input, they are assumed to have the same sampling rate. If \code{P} and \code{X} have different sampling rates, then this input can have two elements (first for \code{P}, second for \code{X}).
#' @param prop The proportion of the maximal excursion to use for defining the "destination" phase of a dive or flight. For example, if \code{prop} is 0.85 (the default), then the destination phase lasts from the first to the last time depth/altitude exceeds 0.85 times the within-dive maximum.
#' @param angular Is X angular data? Defaults to FALSE. 
#' @param X_name A short name to use for X variable in the output data frame. For example, if X is pitch data, use X_name='pitch' to get outputs column names like mean_pitch, etc. Defaults to 'angle' for angular data and 'aux' for non-angular data.
#' @param na.rm Logical, default is TRUE. If TRUE, then returned mean values ignore missing values, computing an average over all non-missing observations.
#' @export
#' @return A data frame with one row for each dive/flight and columns as detailed below. All times are in seconds, and rates in units of x/sec where x is the units of \code{P}.
#' \itemize{
#' \code{max} { The maximum depth or altitude}
#' \code{st} {start time of dive (seconds) - from input dive_cues}
#' \code{et} {end time of dive (seconds) - from input dive_cues}
#' \code{dur} { The duration of the excursion}
#' \code{dest_st} { The start time of the destination phase in seconds since start of tag recording (which is also the end time of to phase)}
#' \code{dest_et} { The end time of the destination phase in seconds since start of tag recording (which is also the start of the from phase).}
#' \code{dest_dur} { The duration in seconds of destination phase}
#' \code{to_dur} { The duration in seconds of to phase}
#' \code{from_dur} { The duration in seconds of from phase}
#' \code{mean_angle} { If angular=TRUE and X is input, the mean angle for the entire excursion. Values for each phase are also provided in columns \code{mean_to_angle}, \code{mean_dest_angle}, and \code{mean_from_angle}.}
#' \code{angle_var} { If angular=TRUE and X is input, the angular variance for the entire excursion. Values for each phase are also provided individually in columns \code{to_angle_var}, \code{dest_angle_var}, and \code{from_angle_var}.}
#' \code{mean_aux} { If angular=FALSE and X is input, the mean value of X for the entire excursion. Values for each phase are also provided in columns \code{mean_to_aux}, \code{mean_dest_aux}, and \code{mean_from_aux}.}
#' \code{aux_sd} { If angular=FALSE and X is input, the standard deviation of X for the entire excursion. Values for each phase are also provided individually in columns \code{to_aux_sd}, \code{dest_aux_sd}, and \code{from_aux_sd}.}
#'#' }
#' @seealso \code{\link{find_dives}}

dive_stats <- function(P, X = NULL, dive_cues, sampling_rate = NULL, 
                       prop = 0.85, angular = FALSE, X_name = NULL,
                       na.rm = TRUE){
  if (!is.list(P) & missing(sampling_rate)){
    stop('For vector input data, sampling_rate must be provided')
  }
  
  if (is.null(X_name)){
    if (angular){
      X_name = 'angle'
    }else{
      X_name = 'aux'
    }
  }
  
  if (is.list(P) & hasName(P, 'data') & hasName(P, 'sampling_rate')){
    fs <- P$sampling_rate
    P <- as.matrix(P$data, ncol=1)
  } else{
    fs <- unlist(head(sampling_rate,1))
  }
  
  if (is.list(X) & hasName(X, 'data') & hasName(X, 'sampling_rate')){
    if (X$sampling_rate != fs){
      xfs <- X$sampling_rate
      #stop('Sampling rates of P and X must match.')
    } else {
      xfs <- fs
    }
    X <- as.matrix(X$data, ncol=1)
  } else {
    if (!is.null(X)){
      X <- as.matrix(X, ncol=1)
      xfs <- unlist(tail(sampling_rate,1))
      if (xfs == fs & nrow(X) != nrow(P)){
        warning('inputs P and X have different numbers of observations, but the same sampling rate. Use sampling_rate input if you need to specify different sampling rates for each one.')
      }
    }
  }
  
  di <- round(dive_cues*fs)
  
  Y <- data.frame(num=c(1:nrow(dive_cues)))
  
  for (d in 1:nrow(dive_cues)){#loop over dives
    z <- P[di[d,1]:di[d,2]] 
    Y$max[d] <- max(z, na.rm=TRUE)
    pt <- range(which(z > prop * max(z)), na.rm=TRUE)
    Y$dur[d] <- dive_cues[d,2] - dive_cues[d,1]
    Y$dest_st[d] <- pt[1]/fs + dive_cues[d,1]
    Y$dest_et[d] <- pt[2]/fs + dive_cues[d,1]
    Y$dest_dur[d] <- Y$dest_et[d] - Y$dest_st[d]
    Y$to_dur[d] <- pt[1]/fs
    Y$to_rate[d] <- (z[pt[1]] - z[1])/Y$to_dur[d]
    Y$from_dur[d] <- (1/fs)*(length(z)-pt[2])
    Y$from_rate[d] <- (utils::tail(z,1) - z[pt[2]])/Y$from_dur[d]
    if (!is.null(X)){
      if (xfs != fs){
        dix <- round(dive_cues*xfs)
        ptx <- round(pt / fs * xfs)
      } else{
        dix <- di
        ptx <- pt
      }
      if (angular){#angular data
        a <- X[dix[d,1]:dix[d,2]] 
        at <- a[c(1:ptx[1])]
        af <- a[c(ptx[2]:length(a))]
        ad <- a[c(ptx[1]:ptx[2])]
        if (na.rm){
          # CircStats functions return NA if any NAs present
          # (no na.omit input avail)
          a <- na.omit(a)
          at <- na.omit(at)
          af <- na.omit(af)
          ad <- na.omit(ad)
        }
        Y$mean_angle[d] <- CircStats::circ.mean(a)
        Y$angle_var[d] <- CircStats::circ.disp(a)$var
        Y$mean_to_angle[d] <- CircStats::circ.mean(at)
        Y$mean_dest_angle[d] <- CircStats::circ.mean(ad)
        Y$mean_from_angle[d] <- CircStats::circ.mean(af)
        Y$to_angle_var[d] <- CircStats::circ.disp(at)$var
        Y$dest_angle_var[d] <- CircStats::circ.disp(ad)$var
        Y$from_angle_var[d] <- CircStats::circ.disp(af)$var
      }else{
        #not angular data
        a <- X[dix[d,1]:dix[d,2]] 
        at <- a[c(1:ptx[1])]
        af <- a[c(ptx[2]:length(a))]
        ad <- a[c(ptx[1]:ptx[2])]
        Y$mean_aux[d] <- mean(a, na.rm=TRUE)
        Y$aux_sd[d] <- stats::sd(a, na.rm=TRUE)
        Y$mean_to_aux[d] <- mean(at, na.rm=TRUE)
        Y$mean_dest_aux[d] <- mean(ad, na.rm=TRUE)
        Y$mean_from_aux[d] <- mean(af, na.rm=TRUE)
        Y$to_aux_sd[d] <- stats::sd(at, na.rm=TRUE)
        Y$dest_aux_sd[d] <- stats::sd(ad, na.rm=TRUE)
        Y$from_aux_sd[d] <- stats::sd(af, na.rm=TRUE)
      }
    }#end processing X
  }#end loop over dives
  
  #change output column names if needed
  if (!(X_name %in% c('angle', 'aux'))){
    names(Y) <- gsub(pattern='angle', replacement=X_name, x=names(Y))
    names(Y) <- gsub(pattern='aux', replacement=X_name, x=names(Y))    
  }
  
  # add in dive start/end times from input dive_cues
  # and start-end times of each phase
  Y$st <- dive_cues[,1]
  Y$et <- dive_cues[,2]
  
  Y <- dplyr::select(Y, num, max, st, et, dur, 
                     dest_st, dest_et, dest_dur,
                     to_dur, from_dur,
                     dplyr::everything())
  
  return(Y)
  }

File: matlab/stats/dive_stats.m
function Y = dive_stats(P, dive_cues, X, fs, prop, angular, X_name)

% Compute summary statistics for dives or flights, given a depth/altitude profile and a series of dive/flight start and end times
% 
% In addition to the maximum excursion and duration, dive_stats} 
% divides each excursion into three phases:
% "to" (descent for dives, ascent for flights), "from" (ascent for dives, descent for flights),
% and "destination". 
% The "destination" (bottom for dives and top for flights) 
% phase of the excursion is identified using a "proportion of maximum 
% depth/altitude" method, whereby for example the bottom phase of a dive 
% lasts from the first to the last time the depth exceeds a stated 
% proportion of the maximum depth. Average vertical velocity is computed 
% for the to and from phases using a simple method: total depth/altitude 
% change divided by total time. If an angular data variable is also 
% supplied (for example, pitch, roll or heading),
% then the circular mean (computed via circ.mean from the CircStats package, 
% available at 
% https://www.mathworks.com/matlabcentral/fileexchange/10676-circular-statistics-toolbox--directional-statistics- .                          ) and variance (computed via \link(CircStats){circ.disp}} and reporting the var} output)
% are also computed for each dive phase and the dive as a whole.
% 
% Inputs:
% P             Depth data. A vector (or one-column matrix), or a tag sensor data list.
% dive_cues     A two-column data frame or matrix with dive/flight start 
%               times in the first column and dive/flight end times in the second. May be obtained from \link{find_dives}}. Units should be seconds since start of tag recording.
% X             (optional) Another data stream (as a vector 
%               (or a one-column matrix) or a tag sensor data list) for 
%               which to compute mean and variability. If angular is 1, 
%               interpreted as angular data (for example pitch, roll, or 
%               heading) and means and variances are computed accordingly.
%               The unit of measure must be radians (NOT degrees).
% fs            (optional if P is a tag sensor data list) Sampling rate of 
%               P (and X, if X is given).
% prop          The proportion of the maximal excursion to use for defining
%               the "destination" phase of a dive or flight. For example, 
%               if prop is 0.85 (the default), then the destination phase 
%               lasts from the first to the last time depth/altitude 
%               exceeds 0.85 times the within-dive maximum.
% angular       Is X angular data? Defaults to 0. 
% X_name        A short name to use for the X variable in the output data 
%               structure. For example, if X is pitch data, 
%               use X_name='pitch' to get outputs column names like 
%               mean_pitch, etc. Defaults to 'angle' for angular data and 
%               'aux' for non-angular data.
% 
% Returns:
% dt, A structure in which each element is a column vector with one row for each 
% dive/flight and columns as detailed below. All times are in seconds, 
% and rates in units of x/sec where x is the units of P. Elements include:
% max       The maximum depth or altitude
% dur       The duration of the excursion
% dest_st  The start time of the destination phase
% dest_et  The end time of the destination phase
% dest_dur  The duration in seconds of destination phase
% to_st  The start time of the to phase
% to_et  The end time of the to phase
% to_dur  The duration in seconds of to phase
% from_st The start time of the from phase
% from_et The end time of the from phase
% from_dur The duration in seconds of from phase
% mean_angle If angular=1 and X is input, the mean angle for the entire excursion. Values for each phase are also provided in columns mean_to_angle, mean_dest_angle, and mean_from_angle.
% angle_var If angular=1 and X is input, the angular variance for the entire excursion. Values for each phase are also provided individually in columns to_angle_var, dest_angle_var, and from_angle_var.
% mean_aux If angular=0 and X is input, the mean value of X for the entire excursion. Values for each phase are also provided in columns mean_to_aux, mean_dest_aux, and mean_from_aux.
% aux_sd If angular=0 and X is input, the standard deviation of X for the entire excursion. Values for each phase are also provided individually in columns to_aux_sd, dest_aux_sd, and from_aux_sd.
%
% Calling syntax examples:
% Y = dive_stats(P, dive_cues); % depth/altitude data only; data is a
%                                % tag sensor data structure
% Y = dive_stats(P, dive_cues, [], fs); % depth/altitude data only; data is 
%                                        % a vector sampled at fs Hz
% Y = dive_stats(P, dive_cues, X, [],[],1) % two sensor data structures,
%                                           % second one is angular data
%
% Standalone example:
% load_nc('testset7');           
% dive_cues = find_dives(P, 5); % get dive cues for dives deeper than 5 m
% dive_cues_matrix = [dive_cues.start'; dive_cues.end']
% Y = dive_stats(P, dive_cues_matrix'); % need to transpose the matrix
%                                       % since we put it in two rows
%                                       % but we need it in two columns
% Y
% Returns: the summary statistics for the dive
%
% valid: Matlab, Octave
% sld33@calvin.edu
% Last modified: 7 Aug 2017

if isstruct(P)
    p = P.data;
    fs = P.sampling_rate;
end

if nargin < 2
    fprintf('inputs P and dive_cues are required\n');
    help dive_stats
    return
end

if nargin < 3 
    X = [];
end

if isstruct(X)
    if X.sampling_rate == P.sampling_rate
       x = X.data;
    else
        fprintf('P and X must be sampled at the same rate\n');
        return
    end
else
    if ~isempty(X) && size(X,1) ~= size(p,1)
       fprintf('P and X must be sampled at the same rate\n');
       return
    else
       x = X;
    end
end

if nargin < 4 && ~isstruct(P)
    fprintf('fs input is required unless P is a sensor data structure\n');
    help dive_stats
    return
end

if nargin < 5 || isempty(prop)
    prop=0.85;
end

if nargin < 6 || isempty(angular)
    angular = 0;
end

if nargin < 7 || isempty(X_name)
    if angular==1
        X_name = 'angle';
    else
        X_name = 'aux';
    end
end

%preallocate space
di = round(dive_cues.*fs);
dur = zeros(size(dive_cues,1),1);
maxz = dur;
dest_st = dur; dest_et = dur; dest_dur = dur;
to_dur = dur; to_rate = dur;
from_dur = dur; from_rate = dur;
if angular == 1
    mean_angle = dur;
    angle_var = dur;
    mean_to_angle = dur;
    mean_dest_angle = dur;
    mean_from_angle = dur;
    to_angle_var = dur;
    dest_angle_var = dur;
    from_angle_var = dur;
else if angular~=1 && ~isempty(x)
    mean_aux=dur;
    aux_sd=dur;
    mean_to_aux=dur;
    mean_from_aux=dur;
    mean_dest_aux=dur;
    from_aux_sd=dur;
    to_aux_sd=dur;
    dest_aux_sd=dur;
end

for d = 1:size(dive_cues,1) %loop over dives
  z = p((di(d,1):di(d,2))) ;
  maxz(d) = nanmax(z);
  zz = find(z > (prop * max(z))) ;
  S = max(zz) ;
  L = min(zz) ;
  dur(d) = dive_cues(d,2) - dive_cues(d,1);
  dest_st(d) = S/fs;
  dest_et(d) = L/fs;
  dest_dur(d) = dest_et(d) - dest_st(d);
  to_dur(d) = S/fs;
  to_rate(d) = (z(S) - z(1))/to_dur(d);
  from_dur(d) = (1/fs)*(length(z)-L);
  from_rate(d) = (z(end) - z(L))/from_dur(d);
  if (~isempty(x))
    if angular==1 
        a = x(di(d,1):di(d,2)); 
        at = a(1:S);
        af = a(L:length(a));
        ad = a(S:L);
        mean_angle(d) = circ_mean(a);
        angle_var(d) = circ_var(a);
        mean_to_angle(d) = circ_mean(at);
        mean_dest_angle(d) = circ_mean(ad);
        mean_from_angle(d) = circ_mean(af);
        to_angle_var(d) = circ_var(at);
        dest_angle_var(d) = circ_var(ad);
        from_angle_var(d) = circ_var(af);
  else
    %not angular data
    a = X(di(d,1):di(d,2)); 
    at = a(1:S);
    af = a(L:end);
    ad = a(S:L);
    mean_aux(d) = nanmean(a);
    aux_sd(d) = std(a(~isnan(a)));
    mean_to_aux(d) = nanmean(at);
    mean_dest_aux(d) = nanmean(ad);
    mean_from_aux(d) = nanmean(af);
    to_aux_sd(d) = std(at(~isnan(at)));
    dest_aux_sd(d) = std(ad(~isnan(ad)));
    from_aux_sd(d) = std(af(~isnan(af)));
    end
  end % end processing X
end %end loop over dives
  %change output column names if needed
  Y = struct('dur', dur, 'maxz', maxz, 'dest_st', dest_st, 'dest_et', ...
      dest_et, 'dest_dur', dest_dur, 'from_dur', from_dur, 'from_rate', ...
      from_rate, 'to_rate', to_rate);
  if angular ~= 1 && ~isempty(x)
  Y.(['mean_' X_name]) = mean_aux;
  Y.(['mean_to_' X_name]) = mean_to_aux;
  Y.(['mean_dest_' X_name]) = mean_dest_aux;
  Y.(['mean_from_' X_name]) = mean_from_aux;
  Y.([X_name '_sd']) = aux_sd;
  Y.(['to_' X_name '_sd' ]) = to_aux_sd;
  Y.(['dest_' X_name '_sd']) = dest_aux_sd;
  Y.(['from_' X_name '_sd']) = from_aux_sd;
  elseif ~isempty(x)
      Y.(['mean_' X_name]) = mean_angle;
  Y.(['mean_to_' X_name]) = mean_to_angle;
  Y.(['mean_dest_' X_name]) = mean_dest_angle;
  Y.(['mean_from_' X_name]) = mean_from_angle;
  Y.([X_name '_var']) = aux_var;
  Y.(['to_' X_name '_var' ]) = to_aux_var;
  Y.(['dest_' X_name '_var']) = dest_aux_var;
  Y.(['from_' X_name '_var']) = from_aux_var;
  end
  return
end
